src/System.Text.Json/tests/WritableJsonApiTests.cs

```diff
@@ -625,5 +660,187 @@ public static void TestAquiringAllProperties()
 
             IEnumerable<JsonNode> fullTimeEmployees = employees.GetAllProperties("FTE");
         }
+
+        /// <summary>
+        /// Transforming JsoneNode to JsonElement
+        /// </summary>
+        [Fact]
+        public static void TestTransformingJsonNodeToJsonElement()
+        {
+            // Send Json through network
+            var employeeDataToSend = EmployeesDatabase.GetNextEmployee().Value;
+            Mailbox.SendEmployeeData(employeeDataToSend.AsJsonElement());
+
+            // Send Json that is lightweight and immutable
+            Mailbox.SendEmployeeData(JsonDocument.Parse(employeeDataToSend).RootElement);
```

Comment:
This doesn't make sense to me.  If SendEmployeeData takes a JsonElement that's a promise to not modify the data; why re-parse it into the immutable model?

If there's a badly-behaving method there's a workaround:

```C#
private static JsonDocument MakeDocument(JsonNode newDocumentRoot)
{
    using (MemoryStream stream = new MemoryStream())
    {
        using (Utf8JsonWriter writer = new Utf8JsonWriter(stream))
        {
            newDocumentRoot.WriteTo(writer);
        }

        writer.Seek(0, SeekOrigin.Beginning);
        return JsonDocument.Parse(stream);
    }
}
```

So I'd recommend leaving this off until a need is demonstrated.

Reply:
So, in my understanding `JsonElement` was supposed to be a type which can hold either `JsonNode` or `JsonDocument` (to be able to have methods that can work on either of them). And that we needed `IsImmutable` property if we wanted to use the information what's really there. So I tried to find scenarios for which there will be a difference and scenarios that will use `IsImmutable`.  I assumed `AsJsonElement` with only box `JsonNode` (`IsImmutable` set to false), while `JsonDocument.Parse` will really read it and transform. With this example I wanted to show the difference of passing just boxed `JsonNode` and the readonly one which I assumed is lighter and faster to use.

Reply:
> And that we needed IsImmutable property if we wanted to use the information what's really there.

`IsImmutable` being false doesn't tell you that the data isn't trustworthy, just like reading out of a List isn't trustworthy.  It just helps explain why the data seems erratic, and that passing it via callbacks or to helper methods _might_ result in it changing.

It's really about

```C#
Assert.Equal(5, arrayNode.Length);
JsonElement arrayElement = arrayNode.AsJsonElement();
Assert.Equals(5, arrayElement.GetArrayLength());
arrayNode.Add(false);
Assert.Equals(6, arrayElement.GetArrayLength());
```

The `arrayElement`'s behavior changed, which isn't "normal" (given the current JsonElement).

A similar example would be

```C#
Assert.Equals(5, arrayElement.GetArrayLength());
SomeFunc(arrayElement);
Assert.Equals(6, arrayElement.GetArrayLength());

void SomeFunc(JsonElement element)
{
    if (JsonNode.TryGetNode(element, out JsonNode node))
    {
        ((JsonArray)node).Add(42);
    }
    else
    {
        Debug.Log("Darn, they gave me an immutable one");
    }
}
```

Reply:
> IsImmutable being false doesn't tell you that the data isn't trustworthy.

I see, thank you for the example. But what would be the scenario of using this property to behave differently based on what it is set to (that is: use it somehow, because maybe it's not needed at all)? (I thought of `TestIsImmutable`, that's why I wanted to add `JsonDocument.Parse(JsonElement)`, but still it doesn't 100% convince me)

> JsonNode.TryGetNode(element, out JsonNode node)

Do you suggest adding `(Try)GetNode`  to `JsonNode` (it seems very useful, and more explaining what JsonElement is)? Maybe `(Try)GetNode` and also `(Try)GetDocument` should be in `JsonElement` ?

> taking JsonElement that's a promise to not modify the data

You are right, it should behave the same way no matter what is underneath, so only read the data, not modify it.  Though, it could not and make any use of checking `IsImmutable` (but I'm not sure if we need this and if it's a good scenario).

But in this example I wanted to focus on a different aspect of `JsonDocument` than being immutable - being efficient, so I thought it might be better to pass really transformed, faster object.

Reply:
> But in this example I wanted to focus on a different aspect of JsonDocument than being immutable - being efficient, so I thought it might be better to pass really transformed, faster object.

But copying it isn't cheap.  So it's a question if is enough going to be read to warrant the cost of transforming it.  (And JsonObject will likely have a faster indexer if asking for a few properties by name...).  It's hard to know when it's worth doing, and that means it's not something that needs to be super visible.

Reply:
> Do you suggest adding (Try)GetNode to JsonNode (it seems very useful, and more explaining what JsonElement is)?

Potentially.  And the value of IsImmutable (false) would be "that method is going to work".

>  Maybe (Try)GetNode and also (Try)GetDocument should be in JsonElement ?

No.  JsonElement only goes down, it never lets you go up.  The "unboxing" of a JsonNode out of a JsonElement is a JsonNode thing, mainly for getting yourself back via a callback.  Real users of JsonElement shouldn't ever see it / want it / need it.

Reply:
I see, thank you for explanations!